src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+          val allFaces = currentFaces ++ updatedFaces
+          // Check for any actual shared faces between blocks - if these exists blocks
+          // should not be merged since there is a real joint seperating the blocks
+          val nonSharedFaces =
+            allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
+              if (!unique.exists(compareSharedFaces(current, _))) {
+                current +: unique
+              } else {
+                unique
+              }
+            }
+          if (allFaces.diff(nonSharedFaces).isEmpty) {
+            Block(origin, nonSharedFaces)
+          }
         }
-        if (right.nonEmpty) {
-          queue.enqueue(right)
+      }
+    }).collect { case blockType: Block => blockType }
+
+    // Remove redundant faces and remove duplicates
+    val nonRedundantBlocks = reconstructedBlocks map { case block @ Block(center, _) =>
+      Block(center, block.nonRedundantFaces)
+    }
+    val joinedBlocks = 
+      nonRedundantBlocks.foldLeft(Seq.empty[Block]) { (unique, current) =>
+        if (!unique.exists(current.approximateEquals(_))) {
+          current +: unique
+        } else {
+          unique
         }
-        median +: loadBalanceIndicesImpl(queue)
       }
+
+    // Divide blocks into groups with and without processor joints remaining
+    val (remainingBlocks, completedBlocks) = joinedBlocks.partition { block =>
+      block.faces.exists(_.processorJoint)
     }
 
-    def removeMedian[A](s: Seq[A]): (A, Seq[A], Seq[A]) = {
-      val medianIndex = s.length / 2
-      (s(medianIndex), s.take(medianIndex), s.drop(medianIndex + 1))
+    if (remainingBlocks.nonEmpty) {
+      // Merged blocks still contain some processor joints
+      mergeBlocks(remainingBlocks ++ processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
+    } else if (processorBlocks.tail.isEmpty) {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
+        if (!unique.exists(current.approximateEquals(_))) {
+          current +: unique
+        } else {
+          unique
+        }
+      }
+    } else {
+      // Proceed to next processor block
+      mergeBlocks(processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
     }
+  }
+
+  /**
+    * Compares two input blocks and determines whether they share a processor face
+    * @param block1 First input block
+    * @param block2 Second input block
+    * @return List of processor faces that are shared by the two blocks. Will be 
+    *         empty if they share no faces
+    */
+  def compareProcessorBlocks(block1: Block, block2: Block): Seq[Face] = {
+    val processorFaces1 = block1.faces.filter { case face => face.processorJoint }
+    val processorFaces2 = block2.faces.filter { case face => face.processorJoint }
+    val faceMatches = 
```

Comment:
Also at Line 220, I think we'll need to keep it as-is since the faces aren't equal. We're looking for normal vectors and distances that are opposite, so sorting won't give the desired effect.
The nested `map` isn't done over all faces, only those containing a processor joint so it isn't quite as bad as we were thinking... Each block should contain only 2 to 3 processor joints at max.